### ISSUE TITLE
Diktat produces wrong `line:column` numbers in the "fix" mode if multiple rules are applied

### ISSUE THREAD

**Issue by 0x6675636b796f75676974687562** [saveourtool/diktat#1538]
Diktat **1.2.3**, Ktlint **0.46.1**.
Consider this code fragment:

```kotlin
@file:Suppress(
    "MISSING_KDOC_TOP_LEVEL",
    "MISSING_KDOC_CLASS_ELEMENTS",
    "EMPTY_BLOCK_STRUCTURE_ERROR",
)

package org.cqfn.diktat

class                                        Example{fun memberFunction(){}}
```

Diktat (e.g. when run from the [command line](https://github.com/saveourtool/benedikt/blob/master/docs/diktat-cli.adoc)) reports the following errors, all on the same line:

```console
$ diktat --relative Example.kt
Example.kt:9:6: [TOO_MANY_CONSECUTIVE_SPACES] too many consecutive spaces: found: 40. need to be: 1 (diktat-ruleset:too-many-spaces)
Example.kt:9:53: [WRONG_WHITESPACE] incorrect usage of whitespaces for code separation: there should be a whitespace before '{' (diktat-ruleset:horizontal-whitespace)
Example.kt:9:54: [BRACES_BLOCK_STRUCTURE_ERROR] braces should follow 1TBS style: incorrect same line after opening brace (diktat-ruleset:block-structure)
Example.kt:9:54: [MISSING_KDOC_ON_FUNCTION] all public, internal and protected functions should have Kdoc with proper tags: memberFunction (cannot be auto-corrected) (diktat-ruleset:kdoc-methods)
Example.kt:9:74: [WRONG_WHITESPACE] incorrect usage of whitespaces for code separation: there should be a whitespace before '{' (diktat-ruleset:horizontal-whitespace)
Example.kt:9:76: [BRACES_BLOCK_STRUCTURE_ERROR] braces should follow 1TBS style: no newline before closing brace (diktat-ruleset:block-structure)
```

Now, when run in the "fix" mode, Diktat will correct all of the above errors except for the `MISSING_KDOC_ON_FUNCTION`:

```console
$ diktat --relative -F Example.kt
Example.kt:9:21: [MISSING_KDOC_ON_FUNCTION] all public, internal and protected functions should have Kdoc with proper tags: memberFunction (cannot be auto-corrected) (diktat-ruleset:kdoc-methods)
```

As you can see, the `line:column` numbers will no longer be `9:54` but `9:21`, because some of the rules have been already applied and some of the errors fixed. Yet, `9:21` is absolutely misleading, because the resulting (formatted) file will look like this:

```kotlin
@file:Suppress(
    "MISSING_KDOC_TOP_LEVEL",
    "MISSING_KDOC_CLASS_ELEMENTS",
    "EMPTY_BLOCK_STRUCTURE_ERROR",
)

package org.cqfn.diktat

class Example {
    fun memberFunction() {}
}
```

If we run Diktat again (now against the reformatted file), there will be `10:5` instead of `9:21`:

```console
$ diktat --relative Example.kt
Example.kt:10:5: [MISSING_KDOC_ON_FUNCTION] all public, internal and protected functions should have Kdoc with proper tags: memberFunction (cannot be auto-corrected) (diktat-ruleset:kdoc-methods)
```

The problem becomes even more evident after the upgrade to Ktlint **0.47.x** ([API Changes & RuleSet providers](https://github.com/pinterest/ktlint/releases/tag/0.47.0)), because it has changed the way AST elements are traversed.

Consider there're multiple rules, [ordered](https://github.com/saveourtool/diktat/blob/master/diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/OrderedRuleSet.kt) in a specific way:

```mermaid
graph LR;
    A-->B;
    B-->C;
```

Assuming there're **multiple** AST syntax nodes **enclosed** in each other:
```mermaid
graph TD;
    File-->Node;
    Node-->Leaf;
```

Ktlint **0.46** used to recursively apply the above rules in the following order:

```mermaid
flowchart TB;
  subgraph File
    direction LR;
    A1["A"]-->B1["B"];
    B1["B"]-->C1["C"];
  end
  subgraph Node
    direction LR;
    A2["A"]-->B2["B"];
    B2["B"]-->C2["C"];
  end
  subgraph Leaf
    direction LR;
    A3["A"]-->B3["B"];
    B3["B"]-->C3["C"];
  end
    File-->Node;
    Node-->Leaf;
```

As of Ktlint **0.47**, the traversal order is different:

```mermaid
flowchart TB;
  subgraph A
    direction LR;
    File1["File"]-->Node1["Node"]
    Node1["Node"]-->Leaf1["Leaf"]
  end
  subgraph B
    direction LR;
    File2["File"]-->Node2["Node"]
    Node2["Node"]-->Leaf2["Leaf"]
  end
  subgraph C
    direction LR;
    File3["File"]-->Node3["Node"]
    Node3["Node"]-->Leaf3["Leaf"]
  end
    A-->B;
    B-->C;
```

In practice, when _Diktat_ is run in the "fix" mode, this change in _Ktlint_ behaviour results in the same error (e.g.: `MISSING_KDOC_ON_FUNCTION`) reported multiple times against different (incorrect) locations within a file.

Probably, we should change our strategy to first running `KtLint.format()` ignoring any errors followed by `KtLint.lint()`, and only report errors produced by the last `lint()` call.

In particular, this blocks saveourtool/benedikt#3.

**Comment by 0x6675636b796f75676974687562** [comment]
@akuleshov7, @nulls, WDYT?

**Comment by nulls** [comment]
@0x6675636b796f75676974687562, agree. We need to split fix and warn parts: we need to fix everything that possible in first run and then warn about left errors
